Lesson 11

ilya-burinskiy opened the pull request.

Commit:
GistQuestionService (using octokit)

Commit:
Gist model

Commit:
Gist link in flash message

Commit:
Updated ru/en locales

Commit:
Admin gists table

VidgarVii commented on app/services/gist_question_service.rb:
```diff
@@ -0,0 +1,41 @@
+
+  def gist_params
+    {
+      description: "A question about #{@test.title} from TestGuru",
```
весь контент должен лежать в файлах локали

VidgarVii commented on app/controllers/test_passages_controller.rb:
```diff
@@ -1,11 +1,26 @@
+  def gist
+    response = GistQuestionService.new(@test_passage.current_question).call
+
+    if response
```
тут нао быть более голословным. response он как правило будет всегда, если есть доступ к сети. Тут вопрос какой будет ответ. И код обработки ответа надо инкапсулировать в сервисе GistQuestionService 
принято использовать такую практику
```
if result.success?
  Gist.create(...)
  flash[:notice] = ..
else
  ...
end
```

VidgarVii commented on app/controllers/test_passages_controller.rb:
```diff
@@ -1,11 +1,26 @@
+
+  def result; end
+
+  def gist
```
так как была создана модель Gist, то по РЕСТ GIST это отдельный ресурс. Для создания гиста в системе следует использовать отдельный контроллер GistsController

VidgarVii commented on app/services/gist_question_service.rb:
```diff
@@ -0,0 +1,41 @@
+require 'octokit'
```
require гемов в RoR происходит автоматически

VidgarVii commented on app/services/gist_question_service.rb:
```diff
@@ -0,0 +1,41 @@
+    response = @client.create_gist(gist_params)
+  rescue Octokit::ClientError
+    nil
+  end
```
успешным вариантом будет возвращать объект, который будет иметь методы success? и gist_url. Ну мб errors еще

VidgarVii commented on app/services/gist_question_service.rb:
```diff
@@ -0,0 +1,41 @@
+    @question = question
+    @test = @question.test
+    configure_endpoint
+    @client = client || Octokit::Client.new(access_token: Rails.application.credentials.github[:access_token])
```
будет лучше, если все эти токены и пути положить в константу в начале класса. Так настройки будут лучше видно
ACCESS_TOKEN = Rails.application.credentials.github[:access_token])

VidgarVii commented on app/controllers/test_passages_controller.rb:
```diff
@@ -1,11 +1,26 @@
+        url: response.html_url,
+        user: @test_passage.user
+      })
+      flash_options = @gist.save ? { notice: "#{t('.success')} #{@gist.url}" } : { notice: t(".failure") }
```
еще вариант инкапсулировать логику записи гиста в БД в тот же сервис.
Тогда контроллер будет чист, что и требуется по его области ответственности

VidgarVii commented on app/views/admin/gists/_gist.html.erb:
```diff
@@ -0,0 +1,7 @@
+<tr>
+  <td><%= link_to gist.question.body[..25], admin_question_path(gist.question) %></td>
```
rails нам предлагает использовать хелпер truncate https://apidock.com/rails/ActionView/Helpers/TextHelper/truncate

VidgarVii commented on app/views/admin/gists/_gist.html.erb:
```diff
@@ -0,0 +1,7 @@
+<tr>
+  <td><%= link_to gist.question.body[..25], admin_question_path(gist.question) %></td>
+  <td>
+    <a href="<%= gist.url %>" rel="nofollow" target="_blank"><%= gist.url %></a>
```
1 = link_to
2 текстом ссылки должен быть хэш (текстовый идентификатор) гиста

VidgarVii commented on app/views/admin/gists/index.html.erb:
```diff
@@ -0,0 +1,14 @@
+    <tr>
+      <th>Question</th>
+      <th>URL</th>
+      <th>User</th>
```
весь контент должен находиться в локале

Commit:
Github client response class

Commit:
Added GistsController

Commit:
Gist hash method

Commit:
GistQuestionService#call returns GitHubClientRespose

Commit:
GistQuestionService internationalisation

Commit:
Gist resource route

Commit:
Gists views update

- Using internatialisation
- Updated paths
- Using truncate method to trucate question body

VidgarVii commented on app/models/gist.rb:
```diff
@@ -0,0 +1,8 @@
+  belongs_to :user
+  belongs_to :question
+
+  def hash
```
по названию метода не понятно о каком хеше идет речь

VidgarVii commented on app/services/gist_question_service.rb:
```diff
@@ -0,0 +1,50 @@
+  private
+
+  def create_gist
+    gist_resource = @client.create_gist(gist_params)
```
gist_resource тут лишний. Он ни как не используется

VidgarVii commented on lib/clients/git_hub_client_response.rb:
```diff
@@ -0,0 +1,16 @@
+    else
+      @success = false
+    end
+  end
```
в конструкторе лучше не использовать ни каких условных операторов. Его задача сформировать объекты 
Успехом мб наличие html_url в респонсе. И отсутствие ошибок. Вот с этим и стоит работать.

VidgarVii commented on app/models/gist.rb:
```diff
@@ -0,0 +1,8 @@
+  belongs_to :question
+
+  def hash
+    URI.parse(url).path.delete('/')
```
чтобы каждый раз не заниматься поиском хеша в урл, можно сразу этот хеш класть в БД

VidgarVii commented on app/services/gist_question_service.rb:
```diff
@@ -0,0 +1,50 @@
+  ACCESS_TOKEN = Rails.application.credentials.github[:access_token]
+  ROOT_ENDPOINT = Rails.application.credentials.github[:root_endpoint]
+
+  def initialize(question, client: nil)
```
client: nil не имеет смысла. Вот question тоже по дефолту нил, пока не придет аргумент.
Для объявления дефолтного аргумента можно использовать такой подход
```
def initialize(question, client: http_client)

```

VidgarVii commented on app/services/gist_question_service.rb:
```diff
@@ -0,0 +1,50 @@
+
+  def configure_endpoint
+    Octokit.configure do |c|
+      c.api_endpoint = ROOT_ENDPOINT
```
мне кажется это лишняя настройка. Мы ведь не разворачиваем свой гитхаб

VidgarVii commented on app/services/gist_question_service.rb:
```diff
@@ -0,0 +1,50 @@
+  def create_gist
+    gist_resource = @client.create_gist(gist_params)
+  rescue Octokit::ClientError
+    nil
```
нил вообще плохая практика, которую приходится затем отлавливать и обрабатывать. Строй логику без возможного нил.
undefined method '' for NilClass самая частая ошибка в программировании

VidgarVii commented on app/services/gist_question_service.rb:
```diff
@@ -0,0 +1,50 @@
+    content = [@question.body]
+    content += @question.answers.pluck(:body)
+    content.join("\n")
+  end
```
как то не затратно по памяти выйдет
лучше так
```
"#{@question.body}\n#{@question.answers.pluck(:body)}"
```
ну или хотябы так
```
[@question.body, *@question.answers.pluck(:body)].join("\n")
```

Commit:
Added hash_from_url column to gists table

Commit:
GitHubClient and GitsQuestionService fixes

VidgarVii commented on lib/clients/git_hub_client_response.rb:
```diff
@@ -0,0 +1,7 @@
+class GitHubClientResponse
```
это же не клиент. обычный сервис объект

VidgarVii replied:
вообще GitHubClientResponse и GistQuestionService следует положить в 1 модуль. Так как решают общую задачу

VidgarVii commented on app/controllers/gists_controller.rb:
```diff
@@ -0,0 +1,28 @@
+        user: @test_passage.user
+      })
+
+      flash_options = { notice: "#{t('.success')} #{@gist.url}" }
```
хорошо бы тут кликабельную ссылку оставить. А не просто текстом. Смотри http://rusrails.ru/rails-internationalization-i18n-api

VidgarVii commented on lib/clients/git_hub_client_response.rb:
```diff
@@ -0,0 +1,7 @@
+  attr_accessor :gist_url, :gist_hash
+
+  def success?
+    !gist_url.nil?
```
gist_url.present?

Commit:
GitHub module

Commit:
Added link to gist in flash msg

Commit:
Passing href to locale

Commit:
Moved github services from lib to app

Commit:
Gist flash url update
